### PR TITLE
[TG-1287] Parsing generic information of base classes

### PR DESCRIPTION
**Pull request by jeannielynnmoulton**
test-gen pointer update PR: https://github.com/diffblue/test-gen/pull/1401

### CODE REVIEW

File: src/java_bytecode/java_types.h
Comment (thk123) [inline src/java_bytecode/java_types.h]: Nit: could this go in a separate commit

File: src/java_bytecode/java_types.h
Comment (thk123) [inline src/java_bytecode/java_types.h]: I don't think this is a useful constructor, a symbol without an identifier seems a bit meaningless, is it required?

File: src/java_bytecode/java_types.h
Comment (thk123) [inline src/java_bytecode/java_types.h]: As it stands, this class doesn't add anything to do the code base, is there a reason to add it?

File: src/java_bytecode/java_types.h
Comment (thk123) [inline src/java_bytecode/java_types.h]: I think there should be a way to use `generic_types().insert` to do what you want here.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: When is this condition false? If never, consider making it a invariant

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: So the signature for `A` contains nothing about `A`, how strange!

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: So this is the generic information about `A` which we're skipping over to read `LGeneric<LInteger;>;`? Consider including this in the example in the above comment

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Do we not need this information in the case? Like if I have:

```java
class B<U> {}
class A<T> extends B<T> {}
```
Do we not need to somehow annotate in the java_symbol_typet that it is an `A::T` not a `B::U`?

**Comment by majakusber** [comment]
This parses the generic information of base classes. For example, for `A extends B<Integer>`, the symbol for the class `A` now shows that the base class is `B` (this functionality was there already) and it has generic type `Integer` (this is the new functionality).

File: unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java]: Consider here using a different letter to the letter that the class for `Generic` uses to make errors more obvious

File: unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java:58]: Some extra tests that might be useful:

- Classes that extend inner classes (maybe not - I guess the parsing for this is already tested)
- Inner classes that extend classes 
(As in both cases we had issues with parsing the signature)

- Classes that are specialised with a generic type (e.g. `extends Gen<ArrayList<Integer>>` (again maybe not so important as the whats after extends should already have this covered by existing tests)

- Probably something with a three hierarchy
- Something that implements an interface
- Something that implements a generic interface (both specialised and specialised)
- Something that implements a couple of interfaces
- Assuming Java allows this (hopefully not) something that just `extends Generirc` 

Sorry :blush:

**Comment by majakusber** [comment]
After an offline discussion with @thk123, this will be redesigned - the new type extends a symbol type and symbol types are meant to refer to a particular symbol in the symbol table with the matching identifier. However, if we have something like `A<T> extends B<T,Integer>` there is no symbol for `B<T,Integer>` and there never will be (symbols are only created for `B<T,U>` and any full specialisation of the class like `B<Integer,Integer>`).

**Review by thk123 — COMMENTED** [review]
Looks good - suggested a few more tests and had an offline discussion about some potential traps when using symbols. Sorry if some of the comments aren't relevant when considering the replacement for `generic_symbol_typet`

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (romainbrenguier) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I'd suggest using a ternary if operator and marking start const

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (romainbrenguier) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: the second argument of `substr` is optional

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (romainbrenguier) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: I would suggest defining a `can_cast_type` function for `java_generic_typet`  (should be the same as `is_java_generic_type`) and using `type_checked_cast`

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (romainbrenguier) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: `gen_base` and `gen_base_type` are const

File: src/java_bytecode/java_types.h
Comment (romainbrenguier) [inline src/java_bytecode/java_types.h]: Why is there a `C` in this identifier? It seems java specific.

File: src/java_bytecode/java_types.h
Comment (romainbrenguier) [inline src/java_bytecode/java_types.h:551]: can you use `static_cast` instead?

File: src/java_bytecode/java_types.h
Comment (romainbrenguier) [inline src/java_bytecode/java_types.h:556]: I was confused by the name of the function `add`, but yes that this is correct...

**Comment by thk123** [comment]
CI failure looks like it might be nonsense so restarted it, is this ready for a re-review @jeannielynnmoulton ?

File: src/java_bytecode/java_types.h
Comment (thomasspriggs) [reply src/java_bytecode/java_types.h]: Looks like this has been removed.

File: src/java_bytecode/java_types.h
Comment (thomasspriggs) [reply src/java_bytecode/java_types.h]: Looks like this has been removed.

File: src/java_bytecode/java_types.h
Comment (thomasspriggs) [reply src/java_bytecode/java_types.h:551]: No, putting a static cast here causes a compile error.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Yes, the signature of a class contains information about its generic parameters, superclass and interfaces but not necessarily the name of the class itself.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Yes we do, this is however done later, as part of `java_bytecode_convert_class::mark_java_implicitly_generic_class_type`

File: src/java_bytecode/java_types.h
Comment (majakusber) [reply src/java_bytecode/java_types.h]: This class has been removed.

File: src/java_bytecode/java_types.h
Comment (majakusber) [reply src/java_bytecode/java_types.h]: See above, the class was removed.

File: unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java
Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java]: The name of the parameter does not matter here as we are not looking at the symbol for `Generic` at all when parsing the class `DerivedGenericUninst`.

Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java]: We are however testing that the parameter of the base is `java::DerivedGenericUninst::T` rather than `java::Generic::T`.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: This is not used anymore.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: We have followed the design used with other types in our code base - having a `is_.._type` and `to_.._type` functions. Would you prefer to have it changed here?

File: src/java_bytecode/java_types.h
Comment (majakusber) [reply src/java_bytecode/java_types.h]: An explanation from @thk123: 
It is meant to mean comment (that is it goes into the comments section of the irep as opposed to the named subtree section). In practise the important part is the string starts with a `#`.

**Comment by majakusber** [comment]
@here This is now ready for re-review. 

In order to fix failing CI, we went back to the original design for the new type for a generic base which is `java_generic_symbol_typet` that extends `symbol_typet` (used for non-generic bases). The advantage is that if the fact that the base is generic is for whatever reason ignored at any point, nothing will break (this works generally for generics in our code as well as in Java itself).

File: src/java_bytecode/java_types.h
Comment (majakusber) [reply src/java_bytecode/java_types.h]: This has shown to be unnecessary (I made the change because my build was complaining, it is not anymore however). So the change was reverted.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Consider using `optionalt<std::string>` rather than the empty string to denote the non-generic case.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Again - I think `optional` would be clearer here.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp:245]: How did this work if `base` was an `exprt` and `type()` was a `typet`?

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Rather than calling the function twice it is probably clearer to store the result. It might also simplify things if you shift the check for `c.signature.has_value()`  inside the extraction method

File: unit/java_bytecode/java_bytecode_parse_generics/InterfaceGeneric.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/InterfaceGeneric.java]: Missing new line

**Review by thk123 — COMMENTED** [review]
Not yet reviewed the unit tests but suggestions on code, though nothing blocking.

File: unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java]: This doesn't seem to be involved?

File: unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/DerivedGenerics.java]: This doesn't appear to be used anywhere

**Review by thk123 — APPROVED** [review]
These unit tests are really good :+1:

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (chrisr-diffblue) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: I'm a bit out of the loop on the generics stuff now :-) so for people like me, it might also be nice if the Doxygen comments for these functions included an example of an input signature string, and an example of what the returned string looks like.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Would it be better to move this `start` variable and its expression outside the `if()` statement and only do the `find` once?

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp:245]: `base` was a `symbol_typet`, only once called `.add_base(base)` it was wrapped into an `exprt`.

**Comment by majakusber** [comment]
@thk123 @chrisr-diffblue Thanks for reviews, I updated the code correspondingly. Will update the test-gen pointer bump shortly https://github.com/diffblue/test-gen/pull/1401.

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Nit: the minus trailing at the end is a little hard to read, I'd prefer pulling `find_closing_semi_colon_for_reference_type(signature.value(), start) ` into a variable if needed to get param on one line

**Review by thk123 — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me too. It would probably be nicer if you can squash the last four commits into their appropriate base commits before merging, just to keep the commit history cleaner, but other than that it seems good to go :-)

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: "returned superclass reference"?

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Use `has_suffix`

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Could the interface also occur earlier? For example, for `class A extends ArrayList<Closable> implements Closable`, would this find the reference to Closable as an argument to ArrayList rather than the interface as desired?

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Suggest parentheses to clarify whether you intend 1 more or 1 less (i.e. a - (b + 1) vs. (a - b) + 1)

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: generic_base? `gen_base` suggests "generated base" to me

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: Ditto

**Review by smowton — COMMENTED** [review]
Mostly looks fine, but please have a look at comment "Could the interface also occur earlier..." before I approve

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Thank you for pointing this out! Yes, the interface can occur earlier as in the example you gave. However, we are only interested in the reference to the interface if it is generic which is ensured by looking for substring with `<` after the name of the interface. In your example there would be no problem. The problem arises in case like this `class A extends B<Interface<String>> implements Interface<Integer>` where the signature is:
`LB<LInterface<Ljava/lang/String;>;>;LInterface<Ljava/lang/Integer;>;`
The function should return reference `LInterface<Ljava/lang/Integer;>;` but this implementation returns `LInterface<Ljava/lang/String;>;`. I've corrected the code and added unit tests for this case.

Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: And this also added unit tests for nested generics in bases which we didn't have before so double thanks :)

**Comment by thk123** [comment]
This requires approval from @smowton due to modification of `irep_ids.def` - I've created a PR to exclude this file from the high risk section of the code owners - #1791

**Review by smowton — APPROVED** [review]

**Pull request merged**
